connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -1271,10 +1268,35 @@ class Dataset[T] private[sql] (
     val colNames: Seq[String] = col1 +: cols
     new RelationalGroupedDataset(
       toDF(),
-      colNames.map(colName => Column(colName).expr),
+      colNames.map(colName => Column(colName)),
       proto.Aggregate.GroupType.GROUP_TYPE_GROUPBY)
   }
 
+  /**
+   * (Scala-specific) Reduces the elements of this Dataset using the specified binary function.
+   * The given `func` must be commutative and associative or the result may be non-deterministic.
+   *
+   * @group action
+   * @since 3.5.0
+   */
+  def reduce(func: (T, T) => T): T = {
+    val list = this
+      .groupByKey(UdfUtils.groupAllUnderBoolTrue())(PrimitiveBooleanEncoder)
```

Comment:
Not sure if this is a stellar idea, the problem is that some aggregation implementations are optimized for keyless aggregation. By adding a dummy key (which is very hard to detect due to the use of a udf), we won't be able to use these code paths. Can we try to use the ReduceAggregator directly?

Reply:
The current code path is rdd ops. It does not go via ReduceAggregator.

If we want to go via ReduceAggregator, we need to make the agg to support `RelationalGroupedDataset#agg(TypedColumn)`, which is missing today. That code path will actually ignore Dataset[T] and always use `RowEncoder` instead. Following this logic, the only problem left is how could we tell the TypedCol apart from Col as no more extra info needed to pass to the server?

Suggest leave it as a TODO.

Reply:
SGTM please file a ticket.

Reply:
How about `df.groupBy().as[Unit, T].reduceGroups(func).as[T].head`? That should stop us from submitting an aggregate with a group.